pkg/certificates/certs.go

```diff
@@ -70,36 +71,20 @@ func createCACertTemplate(expirationInterval time.Duration) (*x509.Certificate,
 	return rootCert, nil
 }
 
-// Create cert template that we can use on the server for TLS
-func createServerCertTemplate(expirationInterval time.Duration) (*x509.Certificate, error) {
-	serverCert, err := createCertTemplate(expirationInterval)
-	if err != nil {
-		return nil, err
-	}
-	serverCert.KeyUsage = x509.KeyUsageDigitalSignature
-	serverCert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
-	serverCert.Subject = pkix.Name{
-		Organization: []string{Organization},
-		// Do not use the same value with CA.
-		CommonName: "control-plane-server",
-	}
-	return serverCert, err
-}
-
-// Create cert template that we can use on the client for TLS
-func createClientCertTemplate(expirationInterval time.Duration) (*x509.Certificate, error) {
-	clientCert, err := createCertTemplate(expirationInterval)
+// Create cert template that we can use on the client/server for TLS
+func createClientServerCertTemplate(expirationInterval time.Duration, san string) (*x509.Certificate, error) {
+	cert, err := createCertTemplate(expirationInterval, san)
 	if err != nil {
 		return nil, err
 	}
-	clientCert.KeyUsage = x509.KeyUsageDigitalSignature
-	clientCert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
-	clientCert.Subject = pkix.Name{
+	cert.KeyUsage = x509.KeyUsageDigitalSignature
+	cert.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}
```

Comment:
Sorry I am also not sure why the original author separated the data-plane vs control-plane as you raised the question https://github.com/knative-sandbox/control-protocol/pull/265#issuecomment-1478185451
But if it is no problem for having both `x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth` as security practice, I think it is not necessary for the separation.

Reply:
... thinking about this, I can't see any _special_ risk in a having a certificate with both usages.

My thinking:

* Key Usages instruct the other side (client or server) about whether the cert is intended for that authorization purpose.
* In order to confuse the client / server usage, you would need to do something like initiate a simultaneous TCP handshake (where the victim _thinks_ they are the client, but the remote side _thinks_ they are the server).  This is likely to fall down pretty quickly after the TLS handshake, if not before, because none of our protocols are "palindromic" (you can speak them the same as client or server).
* Additionally, we're only using this for network communication, not for storing records, so it would be hard to "trick" a server into signing something and then putting it in storage on the server's behalf.